[CI] IVFKnnVectorQueryCalibratedMergeTests testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #12389 / almalinux-10_platform-support-unix](https://gradle-enterprise.elastic.co/s/kncp6gfaekwvk)
- [elasticsearch-periodic-platform-support #12389 / almalinux-10_platform-support-unix](https://gradle-enterprise.elastic.co/s/cppdi2czidexc)

**Reproduction Line:**
```
./gradlew ":server:test" --tests "org.elasticsearch.search.vectors.IVFKnnVectorQueryCalibratedMergeTests.testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample" -Dtests.seed=9BB6B5809EAF140D -Dtests.locale=ff-Adlm-GN -Dtests.timezone=Pacific/Bougainville -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.search.vectors.IVFKnnVectorQueryCalibratedMergeTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: a value greater than <10>
     but: <10> was equal to <10>
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample
- [main] 2 consecutive failures in step almalinux-10_platform-support-unix
- [main] 1 consecutive failures in pipeline elasticsearch-periodic-platform-support
- [main] 2 failures in test testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample (0.7% fail rate in 282 executions)
- [main] 2 failures in step almalinux-10_platform-support-unix (66.7% fail rate in 3 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -644,9 +644,6 @@ tests:
 - class: org.elasticsearch.xpack.esql.expression.function.aggregate.PercentileTests
   method: "testAggregateIntermediate {TestCase=<exponential_histogram>, <long> #3}"
   issue: https://github.com/elastic/elasticsearch/issues/151787
-- class: org.elasticsearch.search.vectors.IVFKnnVectorQueryCalibratedMergeTests
-  method: testShardMergeKeepsCandidatesPerMaxPersistedRescoreOversample
-  issue: https://github.com/elastic/elasticsearch/issues/151789
 - class: org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT
   method: test {csv-spec:stats.minOfKeyword}
   issue: https://github.com/elastic/elasticsearch/issues/151790

--- a/server/src/main/java/org/elasticsearch/search/vectors/AbstractIVFKnnVectorQuery.java
+++ b/server/src/main/java/org/elasticsearch/search/vectors/AbstractIVFKnnVectorQuery.java
@@ -11,8 +11,8 @@ package org.elasticsearch.search.vectors;
 
 import org.apache.lucene.index.FieldInfo;
 import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.index.LeafReaderContext;
-import org.apache.lucene.index.SegmentReader;
 import org.apache.lucene.internal.hppc.IntObjectHashMap;
 import org.apache.lucene.search.BooleanClause;
 import org.apache.lucene.search.BooleanQuery;
@@ -30,7 +30,6 @@ import org.apache.lucene.search.TotalHits;
 import org.apache.lucene.search.Weight;
 import org.apache.lucene.search.knn.KnnCollectorManager;
 import org.apache.lucene.search.knn.KnnSearchStrategy;
-import org.elasticsearch.common.lucene.Lucene;
 import org.elasticsearch.common.lucene.search.Queries;
 import org.elasticsearch.index.codec.vectors.cluster.BulkNeighborQueue;
 import org.elasticsearch.index.codec.vectors.diskbbq.IvfQueryConfigResolver;
@@ -140,35 +139,22 @@ abstract class AbstractIVFKnnVectorQuery extends Query implements QueryProfilerP
         List<Callable<TopDocs>> tasks = new ArrayList<>(leafReaderContexts.size());
         float maxRescoreOversampleAcrossLeaves = 1f;
         for (LeafReaderContext context : leafReaderContexts) {
-            SegmentReader segmentReader = Lucene.tryUnwrapSegmentReader(context.reader());
-            if (segmentReader == null) {
-                IVFCollectorManager knnCollectorManagerForSegment = getKnnCollectorManager(
-                    IvfSegmentConfig.leafCollectorBudget(k, maxRescoreOversampleAcrossLeaves),
-                    indexSearcher
-                );
-                tasks.add(() -> searchLeaf(context, filterWeight, knnCollectorManagerForSegment, visitRatio));
-                continue;
+            LeafReader leafReader = context.reader();
+            FieldInfo fieldInfo = leafReader.getFieldInfos().fieldInfo(field);
+            float segmentOversample = 1f;
+            IvfSegmentConfig resolved = null;
+            if (fieldInfo != null) {
+                resolved = ivfQueryConfigResolver.resolve(fieldInfo, leafReader);
+                segmentOversample = resolved.rescoreOversample();
+                maxRescoreOversampleAcrossLeaves = Math.max(maxRescoreOversampleAcrossLeaves, segmentOversample);
             }
-            FieldInfo fieldInfo = segmentReader.getFieldInfos().fieldInfo(field);
-            if (fieldInfo == null) {
-                IVFCollectorManager knnCollectorManagerForSegment = getKnnCollectorManager(
-                    IvfSegmentConfig.leafCollectorBudget(k, maxRescoreOversampleAcrossLeaves),
-                    indexSearcher
-                );
-                tasks.add(() -> searchLeaf(context, filterWeight, knnCollectorManagerForSegment, visitRatio));
-                continue;
-            }
-            IvfSegmentConfig resolved = ivfQueryConfigResolver.resolve(fieldInfo, segmentReader);
-
-            float segmentOversample = resolved.rescoreOversample();
-            maxRescoreOversampleAcrossLeaves = Math.max(maxRescoreOversampleAcrossLeaves, segmentOversample);
 
             IVFCollectorManager knnCollectorManagerForSegment = getKnnCollectorManager(
                 IvfSegmentConfig.leafCollectorBudget(k, segmentOversample),
                 indexSearcher
             );
 
-            if (resolved.usePrecondition()) {
+            if (resolved != null && resolved.usePrecondition()) {
                 preconditionQuery(context);
             }
             tasks.add(() -> searchLeaf(context, filterWeight, knnCollectorManagerForSegment, visitRatio));

